cpp/src/io/parquet/page_enc.cu

```diff
@@ -1429,10 +1597,355 @@ __global__ void __launch_bounds__(1024) gpuGatherPages(EncColumnChunk *chunks, c
+  cudf::detail::device_single_thread(
+    [offset_at_level  = column_offsets.device_ptr(),
+     end_idx_at_level = column_ends.device_ptr(),
+     col              = *d_col] __device__() {
+      auto curr_col           = col;
+      size_type off           = curr_col.offset();
+      size_type end           = off + curr_col.size();
+      size_type level         = 0;
+      offset_at_level[level]  = off;
+      end_idx_at_level[level] = end;
+      ++level;
+      // Apply offset recursively until we get to leaf data
+      while (curr_col.type().id() == type_id::LIST) {
+        off = curr_col.child(lists_column_view::offsets_column_index).element<size_type>(off);
+        end = curr_col.child(lists_column_view::offsets_column_index).element<size_type>(end);
+        offset_at_level[level]  = off;
+        end_idx_at_level[level] = end;
+        ++level;
+        curr_col = curr_col.child(lists_column_view::child_column_index);
+      }
+    },
+    stream);
+
+  column_offsets.device_to_host(stream, true);
+  column_ends.device_to_host(stream, true);
+
+  rmm::device_uvector<uint8_t> rep_level(max_vals_size, stream);
+  rmm::device_uvector<uint8_t> def_level(max_vals_size, stream);
+
+  rmm::device_uvector<uint8_t> temp_rep_vals(max_vals_size, stream);
+  rmm::device_uvector<uint8_t> temp_def_vals(max_vals_size, stream);
+  rmm::device_uvector<size_type> new_offsets(0, stream);
+  size_type curr_rep_values_size = 0;
+  {
+    // At this point, curr_col contains the leaf column. Max nesting level is
+    // nesting_levels.size().
+    size_t level              = nesting_levels.size() - 1;
+    curr_col                  = nesting_levels[level];
+    auto lcv                  = lists_column_view(curr_col);
+    auto offset_size_at_level = column_ends[level] - column_offsets[level] + 1;
+
+    // Get empties at this level
+    rmm::device_uvector<size_type> empties(0, stream);
+    rmm::device_uvector<size_type> empties_idx(0, stream);
+    size_t empties_size;
+    std::tie(empties, empties_idx, empties_size) =
+      get_empties(nesting_levels[level], column_offsets[level], column_ends[level]);
+
+    // Merge empty at deepest parent level with the rep, def level vals at leaf level
+
+    // Zip the input and output value iterators so that merge operation is done only once
+    auto input_parent_zip_it = thrust::make_zip_iterator(thrust::make_tuple(
+      thrust::make_constant_iterator(level),
+      thrust::make_transform_iterator(thrust::make_counting_iterator(0),
+                                      [idx            = empties_idx.data(),
+                                       mask           = lcv.null_mask(),
+                                       curr_def_level = def_at_level[level]] __device__(auto i) {
+                                        return curr_def_level +
+                                               ((mask && bit_is_set(mask, idx[i])) ? 1 : 0);
+                                      })));
```

Comment:
Would be nice to name this iterator object, since it's not obvious what it does.